### PR TITLE
Derive UmbralPrivateKeys from labels

### PR DESCRIPTION
**Pull request by cygnusv**
- Created a new class called `UmbralKeyingMaterial` that handles keying material for Umbral and that allows deterministic derivation of `UmbralPrivateKeys` based on labels. 
- Adds to `UmbralKeyingMaterial` a method to create `UmbralPrivateKeys` from a label (Solves #98), which will be used to support per-label Capsules. 
- Creates an abstract class `AbstractUmbralKeyingMaterial` that generalizes common behavior of `UmbralKeyingMaterial` and `UmbralPrivateKey` (i.e., secure serialization and deserialization)
- Refines the `hash_to_bn` function, which was relying on a try/except pattern for classifying the types of inputs. 
- Tests for key derivation based on labels

### CODE REVIEW

**Comment by jMyles** [comment]
leaky-chimney

File: tests/test_keys/test_umbral_keys.py [hunk 1]
```diff
@@ -9,6 +9,32 @@ def test_gen_key():
     umbral_pub_key = umbral_priv_key.get_pubkey()
     assert type(umbral_pub_key) == keys.UmbralPublicKey
 
+def test_derive_key_from_label():
+    master_secret = b"random master secret"
```
Comment (tuxxy) [inline tests/test_keys/test_umbral_keys.py]: This is a fine test in that it shows a key can be derived from arbitrary data a la `master_secret`.

However, I don't think this test is showing proper or secure behavior. A user might think that they can derive a key from low entropy data such as an alphanumeric password.

I don't think this is the secure/proper way to do this, see the actual derivation function below for an alternative.

File: umbral/keys.py [hunk 2]
```diff
@@ -38,6 +42,26 @@ def gen_key(cls, params: UmbralParameters=None):
         bn_key = BigNum.gen_rand(params.curve)
         return cls(bn_key, params)
 
+    @classmethod
+    def derive_key_from_label(cls, master_secret: bytes, label: bytes,
```
[the remark this answers is not in the document]
Comment (tuxxy) [reply umbral/keys.py]: Another option is to move this `derive_key_from_label` method onto `UmbralPrivateKey`.

This way, a user can load up a "master key" they have stored and derive keys from labels in a more simplified API on the key.

[the remark this answers is not in the document]
Comment (cygnusv) [reply umbral/keys.py]: While I completely agree that we should make this as misuse resistance as possible and the problems associated to an arbitrary bytes input, I think that allowing UmbralPrivateKeys to be used as master secret to other derived keys is relatively dangerous. The problem I see is that if we allow to use the "master" UmbralPrivateKey to be used somewhere (e.g., to create KFrags), then in the event of a leakage or a collusion attack, all the derived keys are compromised too. 

So, some alternatives:
1) Creating a MasterSecretKey class to represent a high-entropy random sequence of bytes, and use this to derive UmbralPrivateKeys

2) Creating a special subclass of UmbralPrivateKey that somehow doesn't allow to be used by Umbral directly

Thoughts?

File: tests/test_keys/test_umbral_keys.py (hunk 1 above)
Comment (cygnusv) [reply tests/test_keys/test_umbral_keys.py]: Agree.

File: umbral/keys.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jMyles) [reply umbral/keys.py]: @cygnusv: This pattern sounds good to me.

However, I have never liked the name "MasterKey" (or in this case, "MasterSecretKey") because, even in a world where so many things are called "keys" even though they haven't even a vaguely coherent metaphor with an unlocking device, this one is particularly stretched.  If it will never be used as a key, how about:

```
class MasterSecret(UmbralPrivateKey):
    ...
```

[the remark this answers is not in the document]
Comment (cygnusv) [reply umbral/keys.py]: Oh yes, you are right about the name. I'm probably too promiscuous with the work "key".
Now that I think of, it probably has more sense that MasterSecret (or Secret or whatever) is a superclass...

[the remark this answers is not in the document]
Comment (jMyles) [reply umbral/keys.py]: Yeah?  But isn't this what we want the implementer not to do?  Lest they re-use a key per @cygnusv's concern above?

[the remark this answers is not in the document]
Comment (cygnusv) [reply umbral/keys.py]: OK, I see there are two approaches here:

1) There is a conversion path between `MasterSecret` and `UmbralPrivateKey` (in one direction, the other, or both). This not only includes explicit methods (e.g., `UmbralPrivateKey.to_master_secret()`) but also inheritance relations (e.g., if `MasterSecret` is subclass of `UmbralPrivateKey`, then trivially a `MasterSecret` can already be used as `UmbralPrivateKey`). The problem with this approach is that if there's a way to use the same secret material as an `UmbralPrivateKey`, then we have the problems I mentioned before in case of key leakage. I don't see a clear advantage of this approach, apart from the fact that the users only need to care about one of the classes. 

2) There is no such conversion path. If we go in this direction, then there is no way for the implementer to reuse secret material between `MasterSecrets` and `UmbralPrivateKeys`. This is probably the safest approach.

[the remark this answers is not in the document]
Comment (jMyles) [reply umbral/keys.py]: I don't see any advantages of the former.  If there aren't any, let's go with the latter.

File: tests/test_keys/test_umbral_keys.py (hunk 1 above)
Comment (michwill) [reply tests/test_keys/test_umbral_keys.py]: What if we do something like:
```
b'very secure master seed approve recall age route small fine gospel depend canvas inherit impulse people'
```
(yes, this was a secure seed until I posted it)

File: umbral/bignum.py [hunk 3]
```diff
@@ -253,11 +254,16 @@ def __hash__(self):
 
 
 def hash_to_bn(crypto_items, params):
+    if not isinstance(crypto_items, list):
+        crypto_items = [crypto_items]
```
Comment (tuxxy) [inline umbral/bignum.py]: We should be using argument packing here.

It would look like this:
`def hash_to_bn(params, *crypto_items)`

This packs all the `crypto_items` into a tuple that you can iterate over. You can call it like this:
`new_bn = hash_to_bn(params, x_1, x_2, x_3, x_4, ...)`

Then you don't have to worry about it being an iterable.

File: umbral/bignum.py [hunk 4]
```diff
@@ -253,11 +254,16 @@ def __hash__(self):
...
+
+    valid_instance_types = (BigNum, umbral.point.Point)
+
     blake2b = hashes.Hash(hashes.BLAKE2b(64), backend=backend)
     for item in crypto_items:
-        try:
+        if isinstance(item, valid_instance_types):
```
Comment (tuxxy) [inline umbral/bignum.py]: The try-block is more pythonic here. Is there any reason why you are wanting to switch it?

File: umbral/keys.py [hunk 5]
```diff
@@ -10,13 +10,78 @@
...
 
+from abc import ABC, abstractmethod
+ 
+class AbstractUmbralKeyingMaterial(ABC):
+
+    @abstractmethod
+    def _get_keying_material(self):
+        pass
+
+    def to_bytes(self, password: bytes=None, _scrypt_cost: int=20):
```
Comment (tuxxy) [inline umbral/keys.py]: Looks like you haven't merged the latest changes where I also pass in an encoder to allow arbitrary encoding of the bytes. See #112

File: umbral/keys.py [hunk 6]
```diff
@@ -10,13 +10,78 @@
...
 
 from umbral.config import default_params
-from umbral.point import Point, BigNum
+from umbral.point import Point
+from umbral.bignum import BigNum, hash_to_bn
 from umbral.params import UmbralParameters
 
+from abc import ABC, abstractmethod
+ 
+class AbstractUmbralKeyingMaterial(ABC):
```
Comment (tuxxy) [inline umbral/keys.py]: I don't think an abstract base class is needed here. `UmbralKeyingMaterial` can be made as its own independent class. I think it's far more readable that way.

File: umbral/keys.py [hunk 7]
```diff
@@ -239,3 +261,61 @@ def __eq__(self, other):
...
+                    salt=salt,
+                    info=b"NuCypherKMS/KeyDerivation/"+label,
+                    backend=default_backend()
+                    )
+
+        bn_key = hash_to_bn(hkdf.derive(self.keying_material), params)
+        return UmbralPrivateKey(bn_key, params)
+
+    @classmethod
+    def from_bytes(cls, key_data: bytes, password: bytes=None, _scrypt_cost: int=20):
```
Comment (tuxxy) [inline umbral/keys.py]: Let's keep `to_bytes` and `from_bytes` in `UmbralPrivateKey` so that it matches with `UmbralPublicKey`.

If we want to make a move to put these in a base class, we need to think more about structuring the rest of the key classes like that. Either way, I think something like this should be in its own independent PR.

File: umbral/keys.py [hunk 8]
```diff
@@ -239,3 +261,61 @@ def __eq__(self, other):
...
+    def __init__(self, keying_material: bytes=None):
+        """
+        Initializes an UmbralKeyingMaterial.
+        """
+        if keying_material:
+            if len(keying_material) < 32:
+                raise ValueError("UmbralKeyingMaterial must have size at least 32 bytes.")
+            self.keying_material = keying_material
+        else:
+            self.keying_material = os.urandom(32)
```
Comment (tuxxy) [inline umbral/keys.py]: I believe that Blake2b can fit 64 bytes of entropy in it. Perhaps we generate 64 bytes from urandom and utilize all the available entropy.

File: umbral/keys.py [hunk 9]
```diff
@@ -10,13 +10,78 @@
...
+            ).derive(password)
+
+            umbral_get_keying_material = SecretBox(key).encrypt(umbral_get_keying_material)
+            umbral_get_keying_material += salt
+
+        encoded_key = base64.urlsafe_b64encode(umbral_get_keying_material)
+        return encoded_key
+
+    @classmethod
+    def _decode_get_keying_material(cls, key_data: bytes,
```
Comment (tuxxy) [inline umbral/keys.py]: This method name is a little funky. I think I prefer simply `_get_keying_material`.

File: umbral/keys.py [hunk 10]
```diff
@@ -239,3 +261,61 @@ def __eq__(self, other):
...
+            if len(keying_material) < 32:
+                raise ValueError("UmbralKeyingMaterial must have size at least 32 bytes.")
+            self.keying_material = keying_material
+        else:
+            self.keying_material = os.urandom(32)
+
+    def _get_keying_material(self):
+        return self.keying_material
+
+    def derive_private_key_from_label(self, label: bytes,
```
Comment (tuxxy) [inline umbral/keys.py]: I think I prefer `derive_privkey_by_label`. We have taken the standard to calling all keys `privkey` or `pubkey` in one word without the underscore.

File: umbral/bignum.py (hunk 3 above)
Comment (tuxxy) [reply umbral/bignum.py]: Perhaps this can also be in a different PR, too?

File: umbral/bignum.py (hunk 4 above)
Comment (cygnusv) [reply umbral/bignum.py]: OK, I'll revert this

File: umbral/keys.py (hunk 10 above)
Comment (cygnusv) [reply umbral/keys.py]: OK, great. I'm all in for reducing verbosity.

File: umbral/bignum.py (hunk 3 above)
Comment (cygnusv) [reply umbral/bignum.py]: OK, I will revert this changes. Argument packing makes sense, I will do that in a different PR

File: umbral/keys.py (hunk 8 above)
Comment (cygnusv) [reply umbral/keys.py]: OK

File: umbral/keys.py (hunk 6 above)
Comment (tuxxy) [reply umbral/keys.py]: Sounds good.

File: tests/test_keys/test_umbral_keys.py [hunk 11]
```diff
@@ -15,6 +15,34 @@ def test_gen_key():
     umbral_pub_key = umbral_priv_key.get_pubkey()
     assert type(umbral_pub_key) == keys.UmbralPublicKey
 
+def test_derive_key_from_label():
+    
+    umbral_keying_material = keys.UmbralKeyingMaterial()
+
+    label = b"my_healthcare_information"
+
+    priv_key1 = umbral_keying_material.derive_privkey_by_label(label)
```
Comment (michwill) [inline tests/test_keys/test_umbral_keys.py]: `derive_privkey_by_label` looks a little lengthy, but I cannot think of a better name for the method easily
